jumanji/environments/packing/knapsack/generator.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+
+import chex
+import jax
+import jax.numpy as jnp
+
+from jumanji.environments.packing.knapsack.types import State
+
+
+class Generator(abc.ABC):
+    """Defines the abstract `Generator` base class. A `Generator` is responsible
+    for generating a problem instance when the environment is reset.
+    """
+
+    def __init__(
+        self,
+        num_items: int,
+        total_budget: float,
+    ):
+        """Abstract class implementing the attribute `num_cities`.
+
+        Args:
+            num_items (int): the number of items in the problem instance.
+            total_budget (float): the total budget of the knapsack.
+        """
+        self.num_items = num_items
+        self.total_budget = total_budget
+
+    @abc.abstractmethod
+    def __call__(self, key: chex.PRNGKey) -> State:
+        """Call method responsible for generating a new state.
+
+        Args:
+            key: jax random key in case stochasticity is used in the instance generation process.
+
+        Returns:
+            A `Knapsack` environment state.
+        """
+
+
+class RandomGenerator(Generator):
```

Comment:
Should we call it `UniformGenerator` as well?

Reply:
I was hesitant here because for TSP and CVRP it is common to have random uniform (RUE) instances, but I have not seen the same for KP, so I am not sure here whether we should call it `UniformGenerator`. But happy to change it if you think it makes sense here.